core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/UstadTorrentManagerImpl.kt

```diff
@@ -0,0 +1,138 @@
+package com.ustadmobile.core.torrent
+
+import com.turn.ttorrent.client.*
+import com.turn.ttorrent.client.storage.FairPieceStorageFactory
+import com.ustadmobile.core.account.Endpoint
+import com.ustadmobile.core.io.ext.UstadFileCollectionStorage.createFileCollectionStorage
+import com.ustadmobile.core.util.DiTag
+import kotlinx.coroutines.*
+import org.kodein.di.DI
+import org.kodein.di.direct
+import org.kodein.di.on
+import org.kodein.di.instance
+import org.kodein.di.instanceOrNull
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.io.File
+
+
+interface CommunicationManagerListener {
+
+    fun onCommunicationManagerChanged(
+            meteredNetwork: Boolean
+    )
+}
+
+class UstadTorrentManagerImpl(
+        val endpoint: Endpoint,
+        override val di: DI
+) : UstadTorrentManager, CommunicationManagerListener {
+
+    private val torrentDir: File by di.on(endpoint).instance(tag = DiTag.TAG_TORRENT_DIR)
+
+    private val containerDir by di.on(endpoint).instance<File>(tag = DiTag.TAG_DEFAULT_CONTAINER_DIR)
+
+    private var containerInfoHashMap = mutableMapOf<Long, String>()
+
+    private var activeTorrentDownloadMap = mutableMapOf<Long, ActiveTorrentInfo>()
+
+ /*   private var containerDownloadPathMap = mutableMapOf<Long, String>()
+
+    private var containerManagerMap = mutableMapOf<Long, TorrentManager>()
+
+    private var containerListenerMap = mutableMapOf<Long, DownloadListenerAdapter>()*/
+
+    // on android, di uses connectionManager and a provider to get the current communicationManager
+    // on jvm, communicationManager is a singleton
+    private val communicationManager: UstadCommunicationManager?
+        get() = di.direct.instanceOrNull()
+
+    // lock add and removeTorrent
+    private val torrentLock = Mutex()
+
+    val pieceStorage: FairPieceStorageFactory by lazy {
+        FairPieceStorageFactory.INSTANCE
+    }
+
+    private fun requireCommunicationManager(): UstadCommunicationManager{
+        return communicationManager ?: throw IllegalStateException("no communicationManager found")
+    }
+
+    override suspend fun startSeeding() {
+        torrentDir.listFiles() { file: File ->
+            file.name.endsWith(".torrent")
+        }?.forEach { torrentFile ->
+            torrentLock.withLock {
+                    addTorrent(torrentFile.nameWithoutExtension.toLong(), null)
+            }
+        }
+    }
+
+    override suspend fun addTorrent(containerUid: Long, downloadPath: String?) {
```

Comment:
One problem with addTorrent and the active map: I don't see anything that is watching for completion. Once the download is completed, it is no longer an "active" torrent that should be added (even if connectivity is metered). 

I would suggest something like this:

```
inner class TorrentCompleteListener(private val containerUid: Long): TorrentListener {
     ... //blank implementations of all functions needed
    fun onDownloadComplete() {
          //Remove this from the active list
    }

}
```